### PR TITLE
Refactor ParzenEstimator

### PR DESCRIPTION
**Pull request by contramundum53**
## Motivation
`optuna.samplers._tpe.parzen_estimator` had been difficult to read, and we try to make it simpler by splitting it into "probability distributions" part supporting sampling and log pdf function, and "Parzen estimator" part that returns the mixture distribution given the trials.

## Description of the changes
* Extract "probability distribution" part into `optuna.samplers._tpe.probability_distributions`.
* Refactor `optuna.samplers._tpe.parzen_estimator` to be simpler.

The tests are also simplified by splitting those two parts.

### CODE REVIEW

File: optuna/samplers/_tpe/parzen_estimator.py
Comment (HideakiImamura) [inline optuna/samplers/_tpe/parzen_estimator.py:75]: If `parameters.prior_weight` must be provided when `parameters.consider_prior = True`, the following simplification is possible.
```python
prior_weight = 1.0 if len(transformed_observations) == 0 else parameters.prior_weight
if prior_weight is not None:
    weights = np.apend(weights, prior_weight)
```

File: optuna/samplers/_tpe/probability_distributions.py
Comment (HideakiImamura) [inline optuna/samplers/_tpe/probability_distributions.py:53]: This class has 2 methods, so instead of making this a `NamedTuple`, it would be better to make it a regular class.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the great refactoring. Basically, LGTM. I have two comments. PTAL.

By the way, I concern the decrease of the test coverage. Could you try to improve it?

**Comment by HideakiImamura** [comment]
I think we can try the followings, but it is beyond of the scope of this PR. They can be follow-up.
- I think we can use `_SearchSpaceTransform` if we fix it to handle the batch of distributions. 
- It would be great to hide the internal `***Batched***Distributions` from the parzen estimator, but that is debatable.

File: optuna/samplers/_tpe/parzen_estimator.py
Comment (contramundum53) [reply optuna/samplers/_tpe/parzen_estimator.py:75]: When `parameters.prior_weight is not None` and `parameters.consider_prior == False`, this code wrongly adds prior.

Comment (contramundum53) [reply optuna/samplers/_tpe/parzen_estimator.py:75]: Personally, I think we could just set `prior_weight == 0` when `consider_prior == False`. However, that will be another medium-sized refactoring, so let me tackle that in a follow-up.

File: optuna/samplers/_tpe/probability_distributions.py
Comment (contramundum53) [reply optuna/samplers/_tpe/probability_distributions.py:53]: @HideakiImamura Could you explain the reason why a `NamedTuple` with two methods would be bad?
@c-bata and I agreed that using `NamedTuple` would simplify the perceptual burden of reading the code. 
(There is another side effect of using `NamedTuple` that it's easier to compare two `_MixtureOfProductDistribution`.)

**Comment by contramundum53** [comment]
> By the way, I concern the decrease of the test coverage. Could you try to improve it?

It looks like the coverage is decreasing in places other than `parzen_estimator.py`. I'm not sure why this is happening, could you tell me how to fix it?

**Comment by contramundum53** [comment]
I don't think I have removed any of the tests involving those files.

File: optuna/samplers/_tpe/parzen_estimator.py
Comment (HideakiImamura) [reply optuna/samplers/_tpe/parzen_estimator.py:75]: So, how about the following?

```python
prior_weight = 1.0 if len(transformed_observations) == 0 else parameters.prior_weight
if prior_weight is not None and parameters.consider_prior:
    weights = np.apend(weights, prior_weight)
```

File: optuna/samplers/_tpe/probability_distributions.py
Comment (HideakiImamura) [reply optuna/samplers/_tpe/probability_distributions.py:53]: I see. It is important that the `_MixtureOfProductDistribution` is immutable. Thanks. 

At first, a `NamedTuple` with a method seemed unnatural because it is only a named `Tuple`. I thought `NamedTuple` should behave as a container like `Dict`, but it is immutable. However, I think it is good because there is no inconvenience in using this form due to its great practical advantages.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the explanation. I have another comment on tests. Could you take a look?

File: optuna/samplers/_tpe/parzen_estimator.py
Comment (contramundum53) [reply optuna/samplers/_tpe/parzen_estimator.py:75]: This is also wrong because when `len(transformed_observations) == 0` and `parameters.consider_prior == False`, we *should* add prior.

Comment (contramundum53) [reply optuna/samplers/_tpe/parzen_estimator.py:75]: I could have just written 
```
weights = np.append(
                weights,
                [1.0 if len(transformed_observations) == 0 else parameters.prior_weight],
            )
```
but the type checker didn't let me.

Comment (contramundum53) [reply optuna/samplers/_tpe/parzen_estimator.py]: I simplified the logic here. PTAL.

File: tests/samplers_tests/tpe_tests/test_probability_distributions.py
Comment (contramundum53) [reply tests/samplers_tests/tpe_tests/test_probability_distributions.py:11]: Currently we do not have `log=True` in `probability_distributions.py` layer.

**Comment by contramundum53** [comment]
I pulled the latest master and the test coverage was fixed. The test coverage for `study.py` still drops but I don't think this is relevant to this PR.

File: optuna/samplers/_tpe/parzen_estimator.py
Comment (not522) [inline optuna/samplers/_tpe/parzen_estimator.py:76]: How about simplify it like this?

```python
if len(transformed_observations) == 0:
    weights = np.ones(1)
elif parameters.consider_prior:
    weights = np.append(weights, [parameters.consider_prior])
```

**Review by not522 — COMMENTED** [review]
Sorry for my late response. It's almost LGTM. I'm checking this PR does not break compatibility.

File: optuna/samplers/_tpe/parzen_estimator.py
Comment (contramundum53) [reply optuna/samplers/_tpe/parzen_estimator.py]: Thanks for your suggestion. Fixed.

File: optuna/samplers/_tpe/parzen_estimator.py
Comment (HideakiImamura) [inline optuna/samplers/_tpe/parzen_estimator.py:221]: It would be good to remove this comment or replace with a TODO comment.

File: optuna/samplers/_tpe/parzen_estimator.py
Comment (HideakiImamura) [inline optuna/samplers/_tpe/parzen_estimator.py:248]: Ditto.

File: optuna/samplers/_tpe/probability_distributions.py
Comment (HideakiImamura) [inline optuna/samplers/_tpe/probability_distributions.py:38]: This would be simple name.
```suggestion
_BatchedDistributions = Union[
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the update and answering my questions. Almost LGTM with several minor comments. PTAL.

File: optuna/samplers/_tpe/parzen_estimator.py
Comment (contramundum53) [inline optuna/samplers/_tpe/parzen_estimator.py:221]:
```suggestion
                # TODO(contramundum53): Remove dependency on prior_mu
```

Comment (contramundum53) [reply optuna/samplers/_tpe/parzen_estimator.py]: I applied this change. PTAL.

File: optuna/samplers/_tpe/parzen_estimator.py
Comment (contramundum53) [reply optuna/samplers/_tpe/parzen_estimator.py:248]:
```suggestion
                # TODO(contramundum53): Remove dependency of minsigma on consider_prior.
```

Comment (contramundum53) [reply optuna/samplers/_tpe/parzen_estimator.py]: I applied this change. PTAL.

**Comment by HideakiImamura** [comment]
Thanks for the update. LGTM without CI errors.

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Review by not522 — APPROVED** [review]
LGTM. Sorry for my late response.

**Comment by HideakiImamura** [comment]
@contramundum53 Could you resolve the conflicts? I will merge this PR after your resolving it.

**Comment by c-bata** [comment]
I'm sorry for the delay in my review.

Since this PR has blocked #4326, let me unassign myself and create create follow-up PRs if I have find additional suggestions :bow:

**Comment by c-bata** [comment]
Let me merge this PR since this PR has two approvals.

**Pull request merged**
